### PR TITLE
Move allocator into separate crate

### PR DESCRIPTION
**Pull request by jacob-hughes**
By separating the crates up like this, the allocator can be imported and used in rustgc's libstd without needing to break up the repo or duplicate code.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,16 +7,20 @@ edition = "2018"
...
-rustgc = []
+standalone = ["allocator"]
 
 # Enable various GC based statistics. Stats are disabled by default as they have
 # a run-time cost and are expected to only be used for profiling purposes.
 gc_stats = []
 
 [dependencies]
 libc = "*"
+allocator = { path = "./allocator", optional = true }
```
Comment (ltratt) [inline Cargo.toml]: Is the `./` intentional and/or necessary?

File: allocator/src/lib.rs [hunk 2]
```diff
@@ -173,6 +150,18 @@ impl GcAllocator {
     pub fn init() {
         unsafe { boehm::GC_start_performance_measurement() };
     }
+
+    pub unsafe fn register_thread(stack_base: *mut u8) -> bool {
```
Comment (ltratt) [inline allocator/src/lib.rs:155]: What do the `bool` return types mean? It might be worth briefly documenting them.

File: Cargo.toml (hunk 1 above)
Comment (jacob-hughes) [reply Cargo.toml]: I suppose not. I copied this style from the Rust docs when I reminded myself about how to use workspaces. I can change it if you like.

File: allocator/src/lib.rs (hunk 2 above)
Comment (jacob-hughes) [reply allocator/src/lib.rs:155]: Sure, they indicate whether the thread was successfully (un)/registered. I'll document this.

File: Cargo.toml (hunk 1 above)
Comment (ltratt) [reply Cargo.toml]: I guess it's partly a naive question on my part about what it means to have a sub-crate that's an immediate sub-directory. I haven't really looked at workspaces in any detail, so I don't have a meaningful opinion on this.

File: allocator/src/lib.rs (hunk 2 above)
Comment (jacob-hughes) [reply allocator/src/lib.rs:155]: Fixed in f804f9f

File: Cargo.toml (hunk 1 above)
Comment (jacob-hughes) [reply Cargo.toml]: Since they both seem to work, I've changed it to the more familiar path style that we're used to: 299fb96

**Comment by jacob-hughes** [comment]
Think this one is ready for squashing once you're happy

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Just realised that I needed to update the buildbot script (f33839a761cb1b1ba4b59c47d110f00f0da13963), if you're happy with that I'll squash.

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by jacob-hughes** [comment]
This build failure is quite difficult and it's an ordering problem. It's to do with the fact that our buildbot script uses rustgc. However, this PR makes breaking changes that require changes in rustgc, only when those changes are merged will this buildbot script work. 

The way to fix this is in 3 stages:
1. Apply this patch without running `cargo build +rustgc` in the build.
2. Apply the rustgc PR
3. Raise a PR for fixing the buildbot script again in libgc.

Does this sound ok?

**Comment by ltratt** [comment]
Do the necessary rustgc change stand alone? If so, can we just do a rustgc PR?

**Comment by jacob-hughes** [comment]
Sorry, I missed your comment in the noise. 

They don't I'm afraid. This needs merging for the rustgc changes to make sense.

**Comment by ltratt** [comment]
OK, then, you should tell me what I need to do in simple easy steps. Pretend that I'm an idiot :p

**Comment by jacob-hughes** [comment]
Ok. First ask me to squash, then we'll merge this. Once that's done I'll raise a PR for rustgc and then when that gets merged I'll raise another one for libgc again :)

**Comment by ltratt** [comment]
Yes, sir :) Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**